### PR TITLE
New cross-platform PCA9685 driver.

### PR DESCRIPTION
**Pull request by SalimTerryLi**
**Describe problem solved by this pull request**
New cross-platform PCA9685 driver. Able to merge.

**Describe your solution**
Almost full feature implemented.

**Test data / coverage**
I only have a RPi so I test on that and works well.

**Additional context**
Software Reset function cannot be implemented because I2C address 0x00 cannot be reached normally.
External oscillator currently not supported but can be added easily if there is a hardware test condition.
PWM pulse has a fixed offset and I cannot find a way to solve. Maybe it is come from internal oscillator.
All parameters are updated without reboot required.
#13721

### CODE REVIEW

**Comment by SalimTerryLi** [comment]
This one should able to replace the previous one `pca9685`

File: src/drivers/pca9685_pwm/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+############################################################################
+#
+#   Copyright (c) 2015 PX4 Development Team. All rights reserved.
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/CMakeLists.txt]:
```suggestion
#   Copyright (c) 2019 PX4 Development Team. All rights reserved.
```

File: src/drivers/pca9685_pwm/PCA9685.cpp [hunk 2]
```diff
@@ -0,0 +1,292 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2017 PX4 Development Team. All rights reserved.
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.cpp]:
```suggestion
 *   Copyright (c) 2019 PX4 Development Team. All rights reserved.
```

File: src/drivers/pca9685_pwm/PCA9685.cpp [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+int PCA9685::Start()
+{
+	int ret = init();
+	return ret;
+}
+
+int PCA9685::Stop()
+{
+	disAll();
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.cpp]: What does `disAll` mean? I would spell it out `disableAll`.

File: src/drivers/pca9685_pwm/PCA9685.cpp [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+	setPWM(num_outputs, out);
+
+	return 0;
+}
+
+int PCA9685::setFreq(int freq)
+{
+	uint16_t _output;
+	_output = floorl((double)PCA9685_CLOCK_FREQ / freq);
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.cpp]: You can put that on one line:

```suggestion
	uint16_t _output = floorl((double)PCA9685_CLOCK_FREQ / freq);
```

Also, the leading underscore is usually used for class variables and not local variables.

File: src/drivers/pca9685_pwm/PCA9685.cpp [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+	if (_output > 0x00FF) { // out of divider
+		PX4_ERR("frequency too low");
+		return -EINVAL;
+	}
+
+	float freq_err = ((float)PCA9685_CLOCK_FREQ / (float)(_output + (uint16_t)1)
+			  - (float)(freq * PCA9685_PWM_RES))
+			 / (float)(freq * PCA9685_PWM_RES); // actually asked for (freq * PCA9685_PWM_RES)
+
+	if (fabs(freq_err) > (double)0.01f) { // TODO decide threshold
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.cpp]: Would be better to do that in `float` not `double`. Can you use `fabsf`?

File: src/drivers/pca9685_pwm/PCA9685.h [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+	int init() override;
+
+protected:
+	int probe() override;
+
+	static const uint8_t DEFAULT_MODE1_CFG = 0x20;
+	static const uint8_t DEFAULT_MODE2_CFG = 0x04;
+
+	double _Freq = PWM_DEFAULT_FREQUENCY;
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.h]:
```suggestion
	float _freq = PWM_DEFAULT_FREQUENCY;
```

I don't think this needs `double`, right?

File: src/drivers/pca9685_pwm/PCA9685.cpp [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+	_Freq = (float)PCA9685_CLOCK_FREQ / (float)(_output + (uint16_t)1) / PCA9685_PWM_RES; // use actual pwm freq instead.
+
+	setDivider(_output);
+
+	return PX4_OK;
+
+}
+
+/* driver implement */
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.cpp]: This comment is not helpful, better just remove it.

File: src/drivers/pca9685_pwm/PCA9685.cpp [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+int PCA9685::probe()
+{
+	return I2C::probe();
+}
+
+void PCA9685::setPWM(uint8_t channel, const uint16_t &value)
+{
+	int ret;
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.cpp]: Move `int` down to where `ret` is needed.

File: src/drivers/pca9685_pwm/PCA9685.cpp [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+	ret = transfer(buf, 5, nullptr, 0);
+
+	if (OK != ret) {
+		PX4_ERR("i2c::transfer returned %d", ret);
+	}
+}
+
+void PCA9685::setPWM(uint8_t channel_count, const uint16_t *value)
+{
+	int ret;
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.cpp]: Same here.

File: src/drivers/pca9685_pwm/PCA9685.h [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+	/*
+	 * set clock divider
+	 * this func has Super Cow Powers
+	 */
+	void setDivider(uint8_t value);
+
+	/*
+	 * turn off oscillator
+	 */
+	void gosleep();
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.h]: I don't understand `gosleep` but I might understand `turnOffOscillator`. Please rename to whatever is actually meant.

File: src/drivers/pca9685_pwm/PCA9685.h [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+	/*
+	 * restart output
+	 */
+	void restartosc();
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/PCA9685.h]: Same here, what does this mean?

File: src/drivers/pca9685_pwm/pca9685_pwm.cpp [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file pca9685_pwm.cpp
+ * A cross-platform driver and wrapper for pca9685.
+ * Designed to support all control-groups by binding to correct mixer files
+ * Support multi-chip concatenation to reach 32 output channels and even
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/pca9685_pwm.cpp]: You write here it supports up to 32 but in the code there is a hard limit at 16. What is true?

File: src/drivers/pca9685_pwm/pca9685_pwm.cpp [hunk 13]
```diff
@@ -0,0 +1,598 @@
...
+
+### Examples
+It is typically started with:
+$ linux_pwm start -a 64 -b 1
+
+Use the `mixer` command to load mixer files.
+`mixer load /dev/pca9685 ROMFS/px4fmu_common/mixers/quad_x.main.mix`
+)DESCR_STR");
+
+    PRINT_MODULE_USAGE_NAME("linux_pwm_output", "driver");
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/pca9685_pwm.cpp]: Is this name correct?

File: src/drivers/pca9685_pwm/pca9685_pwm.cpp [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+    auto *instance = new PWMDriverWrapper();
+
+    if (instance!=nullptr) {
+        _object.store(instance);
+        _task_id = task_id_is_work_queue;
+
+        int ch;
+        opterr = 0; // turn off parse error output
+        optind=1; // reset parse index
+        int _Addr=PCA9685_DEFAULT_ADDRESS;
```
Comment (julianoes) [inline src/drivers/pca9685_pwm/pca9685_pwm.cpp]: No leading underscore `_` for local variables. And variables should all be lower case.

**Comment by SalimTerryLi** [comment]
There are still two things to fix.
1. whether the old `pca9685` should be removed or not?
2. the code that I used for hardware test should be removed.

Then it is able to be merged.

**Comment by SalimTerryLi** [comment]
Sorry. It seems I have merged something wrong which caused a messed history.

**Comment by julianoes** [comment]
You can clean up your commits by doing `git rebase -i <oldest hash to keep>`. You can squash them, edit them, etc.

If this replaces the old driver, I would vote that you remove the old one.

**Comment by SalimTerryLi** [comment]
It seems I should change back to px4_getpot to build on all boards. I will fix it later.

**Pull request closed without merging**